be/src/exec/pipeline/pipeline_driver_queue.cpp

```diff
@@ -74,8 +88,218 @@ StatusOr<DriverRawPtr> QuerySharedDriverQueue::take(size_t* queue_index) {
     return driver_ptr;
 }
 
-SubQuerySharedDriverQueue* QuerySharedDriverQueue::get_sub_queue(size_t index) {
-    return _queues + index;
+size_t QuerySharedDriverQueue::size() {
+    size_t size = 0;
+    for (const auto& sub_queue : _queues) {
+        size += sub_queue.queue.size();
+    }
+    return size;
+}
+
+void QuerySharedDriverQueue::update_statistics(const DriverRawPtr driver) {
+    _queues[driver->get_dispatch_queue_index()].update_accu_time(driver);
+}
+
+void QuerySharedDriverQueueWithoutLock::put_back(const DriverRawPtr driver) {
+    _put_back(driver);
+}
+
+void QuerySharedDriverQueueWithoutLock::put_back(const std::vector<DriverRawPtr>& drivers) {
+    for (auto driver : drivers) {
+        _put_back(driver);
+    }
+}
+
+void QuerySharedDriverQueueWithoutLock::put_back_from_dispatcher(const DriverRawPtr driver) {
+    // QuerySharedDriverQueueWithoutLock::put_back_from_dispatcher is identical to put_back.
+    put_back(driver);
+}
+
+void QuerySharedDriverQueueWithoutLock::put_back_from_dispatcher(const std::vector<DriverRawPtr>& drivers) {
+    // QuerySharedDriverQueueWithoutLock::put_back_from_dispatcher is identical to put_back.
+    put_back(drivers);
+}
+
+StatusOr<DriverRawPtr> QuerySharedDriverQueueWithoutLock::take(int dispatcher_id) {
+    // -1 means no candidates; else has candidate.
+    int queue_idx = -1;
+    double target_accu_time = 0;
+    DriverRawPtr driver_ptr;
+
+    for (int i = 0; i < QUEUE_SIZE; ++i) {
+        // we just search for queue has element
+        if (!_queues[i].queue.empty()) {
+            double local_target_time = _queues[i].accu_time_after_divisor();
+            // if this is first queue that has element, we select it;
+            // else we choose queue that the execution time is less sufficient,
```

Comment:
Maybe this explanation can be simplify. E.g. find the queue with smallest execution time (placed before for loop)

Reply:
OK, although this comment has already existed before this PR in `QuerySharedDriverQueue`.